lib/msprime.c

```diff
@@ -3925,11 +3946,21 @@ msp_dtwf_generation(msp_t *self)
             }
             // Merge segments in each parental chromosome
             for (i = 0; i < 2; i ++) {
-                for (node = Q[i].head; node != NULL; node = node->next) {
-                    msp_remove_individual(self, (segment_t *) node->item);
+                segments_to_merge = avl_count(&Q[i]);
+                if (segments_to_merge == 1) {
+                    msp_priority_queue_pop(self, &Q[i]);
+                } else if (segments_to_merge >= 2) {
+                    msp_remove_individuals_from_population(self, &Q[i]);
+                    if (segments_to_merge == 2) {
+                        ret = msp_merge_two_ancestors(self,
+                                (population_id_t) j, label,
+                                msp_priority_queue_pop(self, &Q[i]),
```

Comment:
The order in which arguments are evaluated [isn't defined in C](https://stackoverflow.com/questions/376278/parameter-evaluation-order-before-a-function-calling-in-c), which could lead to non-determinism in the output. Let's assign the result of pop to local variables before calling the func.

Reply:
😮 ok, sounds good.